inc/Engine/Common/JobManager/APIHandler/AbstractSafeAPIClient.php

```diff
@@ -0,0 +1,149 @@
+	 * @param bool   $safe Send safe request WP functions or not, default to not.
+	 * @return mixed The response from the API, or WP_Error if a timeout is active.
+	 */
+	private function send_request( $method, $params = [], $safe = false ) {
+		$api_url = $this->get_api_url();
+
+		if ( get_transient( $this->get_transient_key() . '_timeout_active' ) ) {
+			return new WP_Error( 429, __( 'Too many requests.', 'rocket' ) );
+		}
+
+		$params['method'] = strtoupper( $method );
+
+		$response = $this->send_remote_request( $api_url, $method, $params, $safe );
+
+		if ( is_wp_error( $response ) ) {
+			$this->set_timeout_transients();
+			return $response;
+		}
+
+		$body = wp_remote_retrieve_body( $response );
+		if ( empty( $body ) || ( is_array( $response ) && ! empty( $response['response']['code'] ) && 200 !== $response['response']['code'] ) ) {
+			$this->set_timeout_transients();
+			return new WP_Error( 500, __( 'Not valid response.', 'rocket' ) );
+		}
+
+		$this->delete_timeout_transients();
+
+		return $response;
+	}
+
+	/**
+	 * Set the timeout transients.
+	 */
+	private function set_timeout_transients() {
+		$transient_key = $this->get_transient_key();
+
+		$timeout = (int) get_transient( $transient_key . '_timeout' );
+		$timeout = ( 0 === $timeout )
+			? 300
+			: ( 2 * $timeout <= DAY_IN_SECONDS
+				? 2 * $timeout
+				: DAY_IN_SECONDS
+			);
+
+		set_transient( $transient_key . '_timeout', $timeout, WEEK_IN_SECONDS );
+		set_transient( $transient_key . '_timeout_active', true, $timeout );
+	}
+
+	/**
+	 * Delete the timeout transients.
+	 *
+	 * This method deletes the timeout transients for the API requests. It uses the transient key obtained from the `get_transient_key` method.
+	 * The transients deleted are:
+	 * - `{transient_key}_timeout_active`: This transient indicates if a timeout is currently active.
+	 * - `{transient_key}_timeout`: This transient stores the timeout duration.
+	 *
+	 * @return void
+	 */
+	private function delete_timeout_transients() {
+		$transient_key = $this->get_transient_key();
+		delete_transient( $transient_key . '_timeout_active' );
+		delete_transient( $transient_key . '_timeout' );
+	}
+
+	/**
+	 * Decide which WP core function will be used to send the request based on the params.
+	 *
+	 * @param string $api_url API Url.
+	 * @param string $method Request method (GET or POST).
+	 * @param array  $params Parameters being sent with the request.
+	 * @param bool   $safe Send safe request WP functions or not, default to not.
+	 * @return array|WP_Error
+	 */
+	private function send_remote_request( $api_url, $method, $params, $safe ) {
+		if ( ! $safe ) {
+			return wp_remote_request( $api_url, $params );
+		}
+
+		unset( $params['method'] );
+
+		switch ( $method ) {
+			case 'GET':
+				return wp_safe_remote_get( $api_url, $params );
+			case 'POST':
+				return wp_safe_remote_post( $api_url, $params );
+		}
+
+		return new WP_Error( 400, __( 'Not valid request type.', 'rocket' ) );
```

Comment:
Can the $method value be included in the message? It can ease debugging. Or add a log in debug mode?

Reply:
To do later on